h3.go

```diff
@@ -287,7 +301,92 @@ func Uncompact(in []H3Index, res int) []H3Index {
 
 // --- REGIONS ---
 
-// TODO(gilley) solve nested c struct problem for Polyfill funcs
+// Convert slice of geocoordinates to an array of C geocoordinates (represented in C-style as a
+// pointer to the first item in the array). The caller must free the returned pointer when
+// finished with it.
+func geoCoordsToC(coords []GeoCoord) *C.GeoCoord {
+	if len(coords) == 0 {
+		return nil
+	}
+
+	// Use malloc to construct a C-style struct array for the output
+	cverts := C.malloc(C.size_t(C.sizeof_GeoCoord * len(coords)))
+	pv := cverts
+	for _, gc := range coords {
+		*((*C.GeoCoord)(pv)) = gc.toC()
+		pv = unsafe.Pointer(uintptr(pv) + C.sizeof_GeoCoord)
+	}
+
+	return (*C.GeoCoord)(cverts)
+}
+
+// Convert geofences (slices of slices of geocoordinates) to C geofences (represented in C-style as
+// a pointer to the first item in the array). The caller must free the returned pointer and any
+// pointer on the verts field when finished using it.
+func geofencesToC(geofences [][]GeoCoord) *C.Geofence {
+	if len(geofences) == 0 {
+		return nil
+	}
+
+	// Use malloc to construct a C-style struct array for the output
+	cgeofences := C.malloc(C.size_t(C.sizeof_Geofence * len(geofences)))
+
+	pcgeofences := cgeofences
+	for _, coords := range geofences {
+		cverts := geoCoordsToC(coords)
+
+		*((*C.Geofence)(pcgeofences)) = C.Geofence{
+			verts:    cverts,
+			numVerts: C.int(len(coords)),
+		}
+		pcgeofences = unsafe.Pointer(uintptr(pcgeofences) + C.sizeof_Geofence)
+	}
+
+	return (*C.Geofence)(cgeofences)
+}
+
+// Convert GeoPolygon struct to C equivalent struct.
+func geoPolygonToC(gp GeoPolygon) C.GeoPolygon {
+	cverts := geoCoordsToC(gp.Geofence)
+	choles := geofencesToC(gp.Holes)
+
+	return C.GeoPolygon{
+		geofence: C.Geofence{
+			numVerts: C.int(len(gp.Geofence)),
+			verts:    cverts,
+		},
+		numHoles: C.int(len(gp.Holes)),
+		holes:    choles,
+	}
+}
+
+// Free pointer values on a C GeoPolygon struct
+func freeCGeoPolygon(cgp *C.GeoPolygon) {
+	C.free(unsafe.Pointer(cgp.geofence.verts))
+	ph := unsafe.Pointer(cgp.holes)
+	for i := C.int(0); i < cgp.numHoles; i++ {
+		C.free(unsafe.Pointer((*C.Geofence)(ph).verts))
+		ph = unsafe.Pointer(uintptr(ph) + uintptr(C.sizeof_Geofence))
+	}
+	C.free(unsafe.Pointer(cgp.holes))
+}
+
+// Polyfill returns the hexagons at the given resolution whose centers are within the
+// geofences given in the GeoPolygon struct.
+func Polyfill(gp GeoPolygon, res int) []H3Index {
+	var cgp C.GeoPolygon
+	defer freeCGeoPolygon(&cgp)
```

Comment:
why not:
```
	cgp := geoPolygonToC(gp)
	defer freeCGeoPolygon(&cgp)
```

Reply:
I was thinking that if `geoPolygonToC` panics, there may be memory already allocated through `C.malloc` assigned on pointers in the struct that need to be freed. That said, the only panic I'd expect to happen would be `C.malloc` failing, or some undiscovered bug.

Reply:
Looking at the code a second time, this doesn't even really matter because assignment can't really happen until all the pieces have been brought together anyways. I'll clean this one up too.